### ISSUE TITLE
implement AI Model Placeholder Replacement for OBS AI Analysis Filter

### ISSUE THREAD

**Issue by chrisbraddock** [chrisbraddock/obs-studio#1]
# Specification: AI Model Placeholder Replacement for OBS AI Analysis Filter

## 1. Introduction

This document outlines the plan for replacing the current placeholder AI processing logic in the `obs-ai-analysis-filter` OBS Studio plugin with a real, pre-trained AI vision model. The goal is to enable meaningful, real-time (or near real-time) analysis of video frames captured by OBS.

## 2. Current Placeholder Implementation

The existing plugin (`plugins/obs-ai-analysis-filter/`) contains:

*   **`placeholder_ai_process_frame(const struct obs_source_frame *frame)` function:**
    *   Takes an `obs_source_frame` as input.
    *   Simulates AI processing by checking frame dimensions and generating dummy data based on the frame's timestamp.
    *   Returns an `ai_processing_result` struct.
*   **`struct ai_processing_result`:**
    ```c
    struct ai_processing_result {
        bool detected_something;
        float confidence;
        int placeholder_value; 
    };
    ```
*   The filter calls this placeholder function, respects a configurable Nth-frame throttling setting, and logs the simulated results, including a mock overlay message.

## 3. Requirements for the Real AI Model

*   **Performance:**
    *   **Target:** Capable of processing frames at a rate suitable for interactive feedback. For object detection, aiming for at least 10-30 FPS on modest consumer GPUs (e.g., NVIDIA GTX 16xx / RTX xx60 series or AMD equivalent) would be ideal for smoother experiences. CPU performance should be usable, especially with throttling (e.g., 5-10 FPS).
    *   Inference time per frame should ideally be under 30-50ms on GPU, and under 100-200ms on CPU for less aggressive throttling.
*   **Input:**
    *   Accepts raw video frame data (e.g., BGRA, RGBA, NV12, I420 common in OBS). Preprocessing (resize, normalization, color space conversion) will be handled within the plugin before feeding to the model.
*   **Output:**
    *   Structured data representing the AI's findings. For object detection, this would include:
        *   Bounding boxes (e.g., `x, y, width, height`).
        *   Class labels/IDs.
        *   Confidence scores for each detection.
*   **Resource Constraints:**
    *   **CPU Usage:** Should be manageable, not overwhelming the system when OBS and potentially a game are running.
    *   **GPU Usage (Optional but Preferred):** If GPU-accelerated, should not saturate the GPU.
    *   **VRAM Usage:** Must be modest, ideally under 1-2 GB for the model itself, to coexist with games and OBS.
    *   **Model Size:** On-disk size should be reasonably small for distribution.
*   **Ease of Integration:**
    *   Must be runnable locally within the plugin.
    *   Preferably usable via ONNX Runtime for cross-platform C++ deployment.
    *   Well-documented preprocessing and postprocessing steps.
    *   Permissive license for distribution if the plugin is shared.

## 4. Candidate Model Recommendations

Based on research, the following models are recommended:

### 4.1. Primary Candidate: YOLOv8 Nano (Object Detection)

*   **Model Name:** YOLOv8n (Nano version)
*   **Source:** Ultralytics (https://github.com/ultralytics/ultralytics)
*   **Description:** The smallest and fastest variant of the YOLOv8 object detection family. It is pre-trained on the COCO dataset.
*   **Why it's a good candidate:**
    *   **Speed:** Excellent. Reported ~105 FPS on a laptop GTX 1060 GPU (implies <10ms/frame). CPU inference via ONNX Runtime is also viable, though slower (~50-100ms depending on CPU).
    *   **Accuracy:** Good mAP for its size on COCO dataset (around 37.3 mAP@0.5:0.95). Suitable for detecting common objects.
    *   **Size:** Very small model file (~6MB for `.pt`, ONNX likely similar). Low VRAM usage.
    *   **Integration:**
        *   Directly exportable to ONNX format.
        *   ONNX Runtime provides C++ API for inference.
        *   Extensive documentation and community examples available for YOLO models.
*   **Expected Input (ONNX):**
    *   Typically a `float32` tensor of shape `[batch_size, 3, height, width]` (e.g., `1, 3, 640, 640`).
    *   Image needs to be letterboxed/padded to maintain aspect ratio when resizing to model input size (e.g., 640x640).
    *   Pixel values normalized (e.g., 0-1 range).
    *   Color order likely BGR (needs verification for specific exported ONNX).
*   **Expected Output (ONNX):**
    *   Usually a single tensor of shape `[batch_size, num_detections, 4_bbox_coords + 1_confidence + num_classes]` (e.g., `[1, 8400, 85]` for COCO's 80 classes if output includes object confidence).
    *   Bounding box coordinates might be `[x_center, y_center, width, height]` or `[x1, y1, x2, y2]`.
    *   Requires post-processing: confidence thresholding, Non-Maximal Suppression (NMS).
*   **Potential Integration Strategy:**
    *   Use ONNX Runtime C++ API.
    *   Implement preprocessing: resize frame with letterboxing, normalize, convert to planar BGR float tensor.
    *   Implement postprocessing: apply NMS, extract high-confidence detections.

### 4.2. Secondary Candidate: MobileNetV2/V3 (Image Classification)

*   **Model Name:** MobileNetV2 or MobileNetV3
*   **Source:** TensorFlow Hub, PyTorch Hub, ONNX Model Zoo, Hugging Face.
*   **Description:** Lightweight, efficient models designed for image classification on mobile and edge devices.
*   **Why it's a good candidate:**
    *   **Speed:** Extremely fast, especially on CPU.
    *   **Size:** Very small.
    *   **Integration:** Widely available in ONNX format. Simpler input/output than detection.
*   **Task Suitability:** If detailed object location is not needed, but rather an overall scene classification (e.g., "is this a game menu?", "is this an outdoor scene?"), this would be faster and simpler. This is a fallback if object detection is too slow or complex for the initial real model step.
*   **Expected Input (ONNX):**
    *   Typically `float32` tensor `[batch_size, 3, height, width]` (e.g., `1, 3, 224, 224`).
    *   Image resized, normalized (ImageNet standards).
*   **Expected Output (ONNX):**
    *   Tensor `[batch_size, num_classes]` with logits or probabilities.

### 5.1. `ai_analysis_filter_data` struct:
```c
#include <onnxruntime_cxx_api.h> // For ONNX Runtime

struct ai_analysis_filter_data {
    obs_source_t *context;
    struct ai_processing_result last_ai_result;
    
    // Throttling
    int process_every_n_frames;
    uint64_t current_frame_count;

    // ONNX Runtime specific
    Ort::Env ort_env;
    Ort::Session *ort_session; // Pointer, as session creation can fail
    Ort::AllocatorWithDefaultOptions allocator;
    std::vector<const char*> input_node_names;
    std::vector<const char*> output_node_names;
    std::vector<int64_t> input_node_dims; // e.g. {1, 3, 640, 640}

    // Model path (can be made configurable later)
    char *model_path_str; 

    // Preprocessing buffers
    std::vector<float> preprocessed_input_tensor;
    uint32_t model_input_width;
    uint32_t model_input_height;

    // Mutex for thread safety if model/session is accessed by UI thread too
    // pthread_mutex_t model_mutex; 
};
5.2. New Functions:
static bool ai_model_load(struct ai_analysis_filter_data *filter_data):
Initializes ort_env.
Creates ort_session from a model file (e.g., yolov8n.onnx bundled with plugin or user-specified path).
Discovers input/output node names and dimensions.
Called from ai_analysis_filter_create.
static void ai_model_unload(struct ai_analysis_filter_data *filter_data):
Releases ort_session and other ONNX resources.
Called from ai_analysis_filter_destroy.
static void preprocess_frame_for_model(struct ai_analysis_filter_data *filter_data, const struct obs_source_frame *frame):
Handles frame resizing (with letterboxing for YOLO), color conversion (e.g., BGRA to planar BGR), normalization, and populates filter_data->preprocessed_input_tensor.
static struct ai_processing_result postprocess_model_output(struct ai_analysis_filter_data *filter_data, const std::vector<Ort::Value> &output_tensors):
Interprets the raw output tensors from ONNX Runtime.
For YOLOv8: applies confidence thresholding, Non-Maximal Suppression (NMS).
Populates and returns the ai_processing_result struct.
5.3. Modify ai_analysis_filter_create / ai_analysis_filter_destroy:
Call ai_model_load in create. If loading fails, handle appropriately (e.g., log error, disable processing).
Call ai_model_unload in destroy.
5.4. Replace placeholder_ai_process_frame:
A new function, say static struct ai_processing_result run_ai_inference(struct ai_analysis_filter_data *filter_data, const struct obs_source_frame *frame), will:
Call preprocess_frame_for_model.
Create input tensors for ONNX Runtime using filter_data->preprocessed_input_tensor.
Run filter_data->ort_session->Run(...).
Call postprocess_model_output with the output tensors.
Return the populated ai_processing_result.
ai_analysis_filter_video will call run_ai_inference instead of placeholder_ai_process_frame.
5.5. struct ai_processing_result update:
// Example for object detection
struct ai_detection {
    float x, y, width, height; // Bounding box (e.g., normalized 0-1)
    float confidence;
    int class_id;
    const char* class_name; // Optional: map ID to name
};

struct ai_processing_result {
    bool detections_available;
    std::vector<struct ai_detection> detections; // Using C-style for now, or plan for C++ types
    // Or if C only:
    // struct ai_detection* detections_ptr;
    // size_t num_detections;
    
    // Optional: processing time for diagnostics
    // float inference_time_ms; 
};
(Note: Using std::vector implies C++ usage. If strict C is required, dynamic arrays and manual memory management for detections would be needed.)

5.6. CMakeLists.txt (plugins/obs-ai-analysis-filter/CMakeLists.txt):
Need to find and link ONNX Runtime library.
This typically involves find_package(ONNXRuntime REQUIRED) or manually setting paths to ONNX Runtime headers and libraries.
target_link_libraries(obs-ai-analysis-filter PRIVATE OBS::libobs onnxruntime).
The ONNX Runtime headers need to be in the include path.
The ONNX model file (.onnx) needs to be bundled with the plugin or its path made configurable.
5.7. Filter Properties (ai_analysis_filter_properties):
Add a file path property for the ONNX model if not bundling.
Add properties for confidence threshold, NMS threshold.
Add property to enable/disable AI processing.
6. Throttling Considerations
The existing "Process Every Nth Frame" setting will directly control how often run_ai_inference is called. This is crucial for managing performance with a real model. Further FPS-based throttling could be added later if Nth frame isn't sufficient.

7. Overlay Considerations
With actual detections (bounding boxes, class names, confidences), the simulated overlay log messages in ai_analysis_filter_video can become much more meaningful: info("OVERLAY_SIM: [Detected: %s (%.2f) at (%.0f,%.0f,%.0f,%.0f)]", class_name, confidence, x, y, w, h); Future work would involve using gs_ functions to draw these directly (if feasible and thread-safe from an async filter) or updating a separate text/graphics source.

8. Next Steps / Future Enhancements
Implement C++ ONNX Runtime wrapper classes for easier management.
Bundle a default ONNX model (e.g., YOLOv8n.onnx) with the plugin.
Implement robust error handling for model loading and inference.
Add more sophisticated UI properties (model selection, thresholds, etc.).
Explore GPU acceleration options within ONNX Runtime (e.g., DirectML, CUDA execution providers) and how to configure them.
True graphical overlay implementation.
This specification provides a roadmap for integrating a real AI model, focusing initially on YOLOv8n with ONNX Runtime.
